enhancements/custom-alert-configuration.md

```diff
@@ -0,0 +1,439 @@
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+// +genclient
+// +k8s:openapi-gen=true
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type AlertExclusionConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec AlertExclusionConfigSpec `json:"spec"`
+}
+
+// AlertExclusionConfigSpec describes the alert exclusion configuration
+type AlertExclusionConfigSpec struct {
+    // subjects to exclude from the alert
+    //
+    // +optional
+    Subjects map[string]string `json:"subjects,omitempty"
+}
+
+```
+
+The owner of the alert would create the AlertExclusionConfig resource and
+manage the subjects. In the case of KubeVirt, this would include determining
+if and in what namespace KubeVirt is installed.
+
+```yaml
+kind: AlertExclusionConfig
+apiVersion: config.openshift.io/v1
+metadata:
+  name: pod-disruption-budget-at-limit
+spec:
+  subjects:
+    'system:serviceaccount:kubevirt:kubevirt-controller': 'enabled'
+```
+
+Then we would introduce a validating admission policy that would allow changes
+to the `alerts.openshift.io/PodDisruptionBudgetAtLimit` label only for these
+subjects.
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: poddisruptionbudget-at-limit
+spec:
+  failurePolicy: Ignore
+  paramKind:
+    apiVersion: config.openshift.io/v1
+    kind: AlertExclusionConfig
+  matchConstraints:
+    resourceRules:
+    - apiGroups:   ["policy"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["poddisruptionbudgets"]
+    objectSelector:
+      matchExpressions:
+        - key: "alerts.openshift.io/PodDisruptionBudgetAtLimit"
+          operator: Exists
+  validations:
+    - expression: "(has(params.spec.subjects) && request.userInfo.username in params.spec.subjects &&  params.spec.subjects[request.userInfo.username] == 'enabled') ? true : !((has(object.metadata.labels) && 'alerts.openshift.io/PodDisruptionBudgetAtLimit' in object.metadata.labels) || (oldObject != null && has(oldObject.metadata.labels) && 'alerts.openshift.io/PodDisruptionBudgetAtLimit' in oldObject.metadata.labels))"
+      message: "Setting alerts.openshift.io/PodDisruptionBudgetAtLimit label is not allowed."
+      reason: Forbidden
+```
+
+And bind the parameters with:
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicyBinding
+metadata:
+  name: poddisruptionbudget-at-limit
+spec:
+  policyName: poddisruptionbudget-at-limit
+  paramRef:
+    name: pod-disruption-budget-at-limit-alert-opt-out
+
+```
+
+Now, we allow only `system:serviceaccount:kubevirt:kubevirt-controller` user
+to change `alerts.openshift.io/PodDisruptionBudgetAtLimit` label on all PDBs.
```

Comment:
It's not clear to me from the current enhancement text who is creating the various objects.  You have "The owner of the alert would create the AlertExclusionConfig resource", so that's pretty clear.  But then the less specific "we" for ValidatingAdmissionPolicy and nothing for ValidatingAdmissionPolicyBinding.  Are those expected to all be the alert owner?  If so, you know the subjects you are authorizing.  Won't you also know the PDBs you are expecting them to manage?  And in that case, I'm not clear on what this enhancement adds that you couldn't deliver more simply by including:

    kube_poddisruptionbudget_status_current_healthy{poddisruptionbudget!~"kubevirt-disruption-budget-.*"}

or similar in the alert expression.  Perhaps relying on PDB patterns or KubeVirt namespace consistency is too brittle, and moving the alerts from [a cluster-version operator manifest][1] into something managed more directly by the operator is too complicated?

If the alert expression needs to be static, can ValidatingAdmissionPolicy be dynamic, to try to limit restrict subject label access to specific PDBs?  But I guess if ValidatingAdmissionPolicy could be managed dynamically as information comes in (like the KubeVirt namespace), then  you could also inline the subjects there, and you wouldn't need to create an AlertExclusionConfig.

But basically I'm trying to wrap my head around why we want to build a multi-resource flexible system, when you also have the much more narrowly scoped quasi-goal of:

> Mainly, KubeVirt should be the only one allowed to disable this alert.

[1]: https://github.com/openshift/cluster-kube-controller-manager-operator/blob/152572d0606b4de816d06ea08b293e866467c94f/manifests/0000_90_kube-controller-manager-operator_05_alerts.yaml#L25-L34

Reply:
Yes, all of these should be managed by the alert owner.

It is scoped mainly to kubevirt at this moment, but it is not unreasonable to expect that similar request will come in the future for other components. There might be other cases where people use PDBs with guard pods and want to prevent drain to a certain set of nodes. For example by having storage or other applications tied to these nodes, which require a special migration.

So, we do not want to hard code kubevirt names or namespace, so we can allow additional cases in the future. 

Reconciling the ValidatingAdmissionPolicy and injecting the users there without a CRD is certainly an option for now.